File: cmd/grunt/send.go
package main

import (
	"fmt"
	"log"
	"os"
	"strings"

	"github.com/spf13/cobra"
	"grunt/client"
)

var sendCmd = &cobra.Command{
	Use:   "send <message>",
	Short: "Send a single message to the grunt server",
	Args:  cobra.ExactArgs(1),
	Run: func(cmd *cobra.Command, args []string) {
		login := os.Getenv("GRUNT_LOGIN")
		if login == "" {
			log.Fatal("GRUNT_LOGIN environment variable not set (expected user:password)")
		}
		parts := strings.SplitN(login, ":", 2)
		if len(parts) != 2 {
			log.Fatal("GRUNT_LOGIN invalid (expected user:password)")
		}
		user, password := parts[0], parts[1]

		inviteCode, _ := cmd.Flags().GetString("invite-code")
		if inviteCode == "" {
			log.Fatal("--invite-code flag is required")
		}

		msgText := args[0]
		serverAddr, _ := cmd.Flags().GetString("server")
		if serverAddr == "" {
			serverAddr = "http://localhost:54765"
		}

		client := client.NewClient(serverAddr, user)
		defer client.Close()

		if err := client.Register(password, inviteCode); err != nil {
			if strings.Contains(err.Error(), "409") {
				log.Print("User already registered")
			} else {
				log.Fatalf("Failed to register: %v", err)
			}
		} else {
			log.Print("Registration successful")
		}

		if err := client.Login(password); err != nil {
			log.Fatalf("Failed to login: %v", err)
		}

		if err := client.Connect(); err != nil {
			log.Fatalf("Failed to connect: %v", err)
		}

		if err := client.SendMessage(msgText); err != nil {
			log.Fatalf("Failed to send message: %v", err)
		}

		fmt.Println("Message sent successfully.")
	},
}

func init() {
	rootCmd.AddCommand(sendCmd)
	sendCmd.Flags().String("server", "http://localhost:54765", "Server address")
	sendCmd.Flags().String("invite-code", "", "Invite code (required)")
}

File: deno-client/send.ts
#!/usr/bin/env -S deno run --allow-env --allow-net

import { parseArgs } from "jsr:@std/cli@^1.0/parse-args";

interface Flags {
  server: string;
  inviteCode: string;
}

function parseFlags(): Flags {
  const args = parseArgs(Deno.args, {
    string: ["server", "invite-code"],
    default: {
      server: "http://localhost:54765",
    },
    alias: {
      s: "server",
      i: "invite-code",
    },
  });

  return {
    server: args.server as string,
    inviteCode: (args["invite-code"] as string) ?? "",
  };
}

async function register(
  server: string,
  user: string,
  password: string,
  inviteCode: string,
): Promise<void> {
  const resp = await fetch(`${server}/api/user`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ user, password, invite_code: inviteCode }),
  });

  if (resp.status === 201) {
    console.log("Registration successful");
  } else if (resp.status === 409) {
    console.log("User already registered");
  } else {
    const body = await resp.text();
    throw new Error(`Registration failed: ${resp.status} ${body}`);
  }
}

async function login(
  server: string,
  user: string,
  password: string,
): Promise<string> {
  const resp = await fetch(`${server}/api/user/login`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ user, password }),
  });

  if (resp.status !== 200) {
    const body = await resp.text();
    throw new Error(`Login failed: ${resp.status} ${body}`);
  }

  const data = await resp.json();
  return data.token;
}

async function sendMessage(
  server: string,
  token: string,
  content: string,
): Promise<void> {
  const resp = await fetch(`${server}/api/chat/message`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${token}`,
    },
    body: JSON.stringify({ content }),
  });

  if (resp.status !== 200) {
    const body = await resp.text();
    throw new Error(`Failed to send message: ${resp.status} ${body}`);
  }

  console.log("Message sent successfully.");
}

async function main() {
  const flags = parseFlags();
  const loginEnv = Deno.env.get("GRUNT_LOGIN");

  if (!loginEnv) {
    console.error("Error: GRUNT_LOGIN environment variable not set (expected user:password)");
    Deno.exit(1);
  }

  const [user, password] = loginEnv.split(":");
  if (!user || !password) {
    console.error("Error: GRUNT_LOGIN invalid (expected user:password)");
    Deno.exit(1);
  }

  if (!flags.inviteCode) {
    console.error("Error: --invite-code flag is required");
    Deno.exit(1);
  }

  const message = Deno.args.find((arg) => !arg.startsWith("-"));
  if (!message) {
    console.error("Usage: send.ts <message> [--server URL] [--invite-code CODE]");
    Deno.exit(1);
  }

  try {
    await register(flags.server, user, password, flags.inviteCode);
    const token = await login(flags.server, user, password);
    await sendMessage(flags.server, token, message);
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    console.error(`Error: ${message}`);
    Deno.exit(1);
  }
}

await main();
